straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,283 @@ def find_n_competing(peaks, windows, fraction):
+                s2_dtype.append(((x + ' of previous s2 peak casting ' + r + ' largest shadow [cm]', sa + 'pre_' + x + '_s2'), np.float32))
+                s2_reordered_dtype.append(((x + ' of previous reordered s2 peak casting ' + r + ' largest shadow [cm]', sa + 'pre_' + x + '_s2_re'), np.float32))
+            s2_corr_dtype.append((('previous ' + r + ' largest s2 shadow with position correlation [PE/ns]', sa + 'shadow_s2_corr'), np.float32))
+            s2_corr_dtype.append((('previous ' + r + ' largest s2 shadow position correlation PDF', sa + 'shadow_s2_prob'), np.float32))
+        dt_dtype = []
+        sum_dtype = []
+        for si, s in zip(['s1', 's2', 'reordered s2'], ['_s1', '_s2', '_s2_re']):
+            dt_dtype.append((('Nearest ' + si + ' delta t', 'near_dt' + s), np.int64))
+            sum_dtype.append((('Sum of ' + si + ' shadow in time window', 'shadow_sum' + s), np.float32))
+        dtype = s1_dtype + s2_dtype + s2_reordered_dtype + s2_corr_dtype + dt_dtype + sum_dtype + strax.time_fields
         return dtype
 
     def compute(self, peaks):
-        roi_shadow = np.zeros(len(peaks), dtype=strax.time_fields)
-        roi_shadow['time'] = peaks['center_time'] - self.time_window_backward
-        roi_shadow['endtime'] = peaks['center_time'] - self.n_drift_time
-
-        mask_pre_s2 = peaks['area'] > self.s2_threshold
-        mask_pre_s2 &= peaks['type'] == 2
-        split_peaks = strax.touching_windows(peaks[mask_pre_s2], roi_shadow)
-        res = np.zeros(len(peaks), self.dtype)
-        res['pre_s2_x'] = np.nan
-        res['pre_s2_y'] = np.nan
-        if len(peaks):
-            self.compute_shadow(peaks, peaks[mask_pre_s2], split_peaks, self.exponent, res)
-
-        res['time'] = peaks['time']
-        res['endtime'] = strax.endtime(peaks)
+        return self.compute_s1_s2(peaks, peaks)
+
+    def compute_s1_s2(self, peaks, current):
+        roi_shadow = np.zeros(len(current), dtype=strax.time_fields)
+        roi_shadow['time'] = current['center_time'] - self.time_window_backward
+        roi_shadow['endtime'] = current['center_time']
+
+        resdict = dict()
+        for stype, key, reorder in zip([2, 2, 1], ['s2_re', 's2', 's1'], [True, False, False]):
+            mask_pre = (peaks['type'] == stype) & (peaks['area'] > self.threshold[key])
+            split_peaks = strax.touching_windows(peaks[mask_pre], roi_shadow)
+            array = np.zeros(len(current), np.dtype(self.sdtype + self.ddtype))
+            for sa in ['', 'alt_']:
+                for x in ['x', 'y']:
+                    array[sa + 'pre_' + x] = np.nan
+                array[sa + 'shadow_dt'] = self.time_window_backward
+                array[sa + 'pre_area'] = self.threshold[key]
+                array[sa + 'shadow'] = array[sa + 'pre_area'] * array[sa + 'shadow_dt'] ** self.exponent
+            array['near_dt'] = self.time_window_backward
+            array['shadow_sum'] = 0
+            if len(current):
```

Comment:
Do you really need this if statement? If current is empty also peaks is empty hence you function gets two empty inputs why bothering in that case with the function at all. You could move this if to the start of your function and just return an empty array of the correct dtype in that case.